pkg/controllers/drift/controller.go

```diff
@@ -0,0 +1,135 @@
+package drift
+
+import (
+	"context"
+	"fmt"
+	"github.com/aws/karpenter-core/pkg/apis/provisioning/v1alpha5"
+	operatorcontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+	"github.com/aws/karpenter/pkg/apis/v1alpha1"
+	"github.com/aws/karpenter/pkg/cloudprovider"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"knative.dev/pkg/logging"
+	"net/http"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	"time"
+)
+
+const controllerName = "drift"
+
+type Drift struct {
+	kubeClient    client.Client
+	cloudProvider *cloudprovider.CloudProvider
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client, cloudProvider *cloudprovider.CloudProvider) *Drift {
+	return &Drift{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+	}
+}
+
+func (d Drift) Reconcile(ctx context.Context, req controllerruntime.Request) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named(controllerName).With("provisioner", req.Name))
+	node := &v1.Node{}
+	if err := d.kubeClient.Get(ctx, req.NamespacedName, node); err != nil {
+		return reconcile.Result{}, nil
+	}
+
+	if _, drifted := node.Annotations[v1alpha5.DriftedAnnotationKey]; drifted {
+		return reconcile.Result{}, nil
+	}
+
+	provisionerName, provisionerExists := node.Labels[v1alpha5.ProvisionerNameLabelKey]
+	if !provisionerExists {
+		//Only karpenter owned Nodes.
+		return reconcile.Result{}, nil
+	}
+
+	provisioner := &v1alpha5.Provisioner{}
+	if err := d.kubeClient.Get(ctx, types.NamespacedName{Name: provisionerName}, provisioner); err != nil {
+		return reconcile.Result{}, nil
+	}
+
+	drifted := d.cloudProvider.IsNodeDrifted(ctx, provisioner, *node)
+	if drifted {
+		stored := node.DeepCopy()
+		node.Annotations[v1alpha5.DriftedAnnotationKey] = "true"
+		if err := d.kubeClient.Patch(ctx, node, client.MergeFrom(stored)); err != nil {
+			return reconcile.Result{}, fmt.Errorf("patching node, %w", err)
+		}
+	}
+
+	return reconcile.Result{RequeueAfter: 30 * time.Minute}, nil
+}
+
+func (d Drift) Builder(ctx context.Context, m manager.Manager) operatorcontroller.Builder {
+	if err := m.GetFieldIndexer().IndexField(ctx, &v1alpha5.Provisioner{}, ".spec.providerRef.name", func(rawObj client.Object) []string {
+		provisioner := rawObj.(*v1alpha5.Provisioner)
+		return []string{provisioner.Spec.ProviderRef.Name}
+	}); err != nil {
+		//todo:Need inputs on this, or this indexer needs to be put somewhere else or removed altogether
+	}
```

Comment:
I'm not sure I understand what you're doing here. Is this another pattern you've seen somewhere else? I think we should be able to do everything within the return statement you have at line 81.

Reply:
So its related to `d.kubeClient.List(ctx, provisioners, client.MatchingFields{".spec.providerRef.name": o.GetName()})` which is down below in `Watches(
			&source.Kind{Type: &v1alpha1.AWSNodeTemplate{}},`, if the field is not indexed, it wont work and returns an error that the index doesn't exist. 

Is there a way we can do away with this index ? We need it to get provisioners that are related to a specific `AwsNodeTemplate`.

Reply:
I think you should be able to do this within the EnqueueRequestsFromMapFunc function and just return an empty array if it's not indexed.

Reply:
I think, `IndexField` is a cost intensive operation and can't be done on each reconcile. I am thinking to handle it differently by only adding a `Watches(
			&source.Kind{Type: &v1alpha1.AWSNodeTemplate{}},`, if the index error is not nil by exposing a builder function. Will put that in the next commit for you to verify.

Reply:
@ellistarn Since Nick is out, can you please check if current implementation is fine ?

Reply:
@njtran You might want to check this when you come back. I think @jonathan-innis mentioned in some other comment that we move this indexField to karpenter-core, but I am not sure where.